### PR TITLE
preventing 'old' startup exceptions from shutting down host

### LINKED ISSUE #4089: A slow startup exception from a JobHost can reset the state of the the WebHost and return constant 503s.
The pattern is:
- ScriptHostA is starting.
- Before ScriptHostA is able to call `StartAsync()` on its underlying `JobHost`, a restart request comes in.
- ScriptHostA is orphaned and ScriptHostB starts up.
- ScriptHostA's startup is canceled and it is disposed. It still calls `StartAsync()` with its disposed services.
- Meanwhile, ScriptHostB starts up happily and is running.
- ScriptHostA eventually throws an `ObjectDisposedException` -- usually something related to the `LoggerFactory`. 
- Because our startup exception handling takes action on the currently-running-host, the exception from ScriptHostA (which is not an active host; ScriptHostB is) causes:
  - The whole service to be set into an `Error` state.
  - The active host (ScriptHostB!) to be orphaned.
  - The restart loop to be canceled.
  - The error log to be written using the active host's logger (so ScriptHostA is telling us it has an error -- but using ScriptHostB's instance id. Talk about "tough to diagnose!"). 
- Now we have no running host and no restart occurring. Any requests to the site cannot find a host so they return 503 forever.

### PR DESCRIPTION
**Pull request by brettsam**
Fixes #4089.

I've also:
- Added additional logging that would have made the investigation into this easier. It was really hard to follow due to the number of log-less code paths we can take.
- Added a `ThrowIfCancelationRequested()` at the point I was seeing this error. We already knew startup was canceled, so there's no need to try to start the host.

### CODE REVIEW

File: test/WebJobs.Script.Tests/WebJobsScriptHostServiceTests.cs [hunk 1]
```diff
@@ -105,10 +117,78 @@ public async Task HostRestart_Specialization_Succeeds()
...
+            hostA.Verify(m => m.StopAsync(It.IsAny<CancellationToken>()), Times.Exactly(2));
+            hostA.Verify(m => m.Dispose(), Times.Exactly(2));
+
+            // We should not be calling Orphan on the good host
+            hostB.Verify(m => m.StopAsync(It.IsAny<CancellationToken>()), Times.Never);
+            hostB.Verify(m => m.Dispose(), Times.Never);
+
+            // The "late" exception from the first host shouldn't bring things down
+            Assert.Equal(ScriptHostState.Running, _hostService.State);
+        }
```
Comment (brettsam) [inline test/WebJobs.Script.Tests/WebJobsScriptHostServiceTests.cs]: Every one of these verification groups failed before my fix.

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs [hunk 2]
```diff
@@ -156,32 +158,49 @@ private async Task StartHostAsync(CancellationToken cancellationToken, int attem
...
             {
+                GetHostLogger(localHost).LogDebug("Host startup was canceled.");
                 throw;
             }
             catch (Exception exc)
             {
-                LastError = exc;
-                State = ScriptHostState.Error;
-                attemptCount++;
+                // Another host has been created before this host
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: doesn't this text belong in the else block below?

Comment (brettsam) [reply src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Yep -- missed while refactoring. Will move.

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs [hunk 3]
```diff
@@ -156,32 +158,49 @@ private async Task StartHostAsync(CancellationToken cancellationToken, int attem
             }
             catch (OperationCanceledException)
             {
+                GetHostLogger(localHost).LogDebug("Host startup was canceled.");
```
Comment (pragnagopa) [inline src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Can log this at info level? It might be worth it atleast initially to see these logs in Kusto to get better understanding for JobHost life cycle.

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs [hunk 4]
```diff
@@ -156,32 +158,49 @@ private async Task StartHostAsync(CancellationToken cancellationToken, int attem
...
-                            }
-                        }, TaskContinuationOptions.ExecuteSynchronously);
+                            t.Exception.Handle(e => true);
+                        }
+                    }, TaskContinuationOptions.ExecuteSynchronously);
 
-                cancellationToken.ThrowIfCancellationRequested();
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    logger.LogDebug($"Cancellation requested. A new host will not be started.");
```
Comment (pragnagopa) [inline src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Same here. LogInfo.

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs (hunk 3 above)
Comment (brettsam) [reply src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Kusto sees debug logs. Writing debug is good for stuff like this bc customers won’t see it by default but we will.

Comment (pragnagopa) [reply src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Thanks. Good to know!

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs [hunk 5]
```diff
@@ -390,10 +390,17 @@ private bool ShutdownHostIfUnhealthy()
...
+                var scriptHost = (ScriptHost)instance?.Services.GetService<ScriptHost>();
+                if (scriptHost != null)
+                {
+                    scriptHost.HostInitializing -= OnHostInitializing;
+                }
+            }
+            catch (ObjectDisposedException)
             {
-                scriptHost.HostInitializing -= OnHostInitializing;
+                // If the instance is already disposed, we cannot access its services.
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Leaving this subscription is concerning. The change improve things, but doesn't resolve the issue of possible leaks.

**Comment by brettsam** [comment]
@mathewc / @pragnagopa -- I found another issue while doing some manual testing before committing.  My tests weren't disposing the services when we called Dispose(), so it hid another scenario where we try to access the services of the disposed host. I've add try/catch around those places so this should be more stable. Wanted to get an 'ack' from you guys before I snuck it in.

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs (hunk 5 above)
Comment (fabiocav) [reply src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Can we at least have an issue tracking this?

Comment (brettsam) [reply src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: The fact that the host is disposed means that we've already gone through this codepath once. So the subscription has already been de-registered. I realize it's not guaranteed, that all disposals happen during our call to Orphan, but in practice it seems like they are.

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs [hunk 6]
```diff
@@ -156,32 +158,49 @@ private async Task StartHostAsync(CancellationToken cancellationToken, int attem
...
             {
+                GetHostLogger(localHost).LogDebug("Host startup was canceled.");
                 throw;
             }
             catch (Exception exc)
             {
-                LastError = exc;
-                State = ScriptHostState.Error;
-                attemptCount++;
+                bool isActiveHost = localHost == _host;
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: For safety/clarity, user ReferenceEquals

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs [hunk 7]
```diff
@@ -277,18 +297,26 @@ private IHost BuildHost(bool skipHostStartup = false, bool skipHostJsonConfigura
             return _scriptHostBuilder.BuildHost(skipHostStartup, skipHostJsonConfiguration);
         }
 
-        private ILogger GetHostLogger()
+        private ILogger GetHostLogger(IHost host)
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: 👍

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs (hunk 5 above)
Comment (fabiocav) [reply src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Just wanted to make sure, as I believe there is a chance that wouldn't be the case and there are things we can do to make this a bit more robust.

Comment (brettsam) [reply src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: Oh, yeah, for sure. This change doesn't make things worse, at least :-) If that thing's disposed, there's no way for us to access the ScriptHost with the way things are currently set up. I've opened an issue to see what we can do there: #4092.

File: src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs (hunk 6 above)
Comment (brettsam) [reply src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs]: 👍
